In [6]:
import pandas as pd

file_path = "../data/raw/online_retail_II.csv"

df = pd.read_csv(file_path)

df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


# 1. Dataset Overview

In this section, we inspect the dataset to understand its structure, dimensions, column types, and overall quality before performing any cleaning or feature engineering.

In [7]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 1067371
Columns: 8


In [8]:
df.columns.tolist()

['Invoice',
 'StockCode',
 'Description',
 'Quantity',
 'InvoiceDate',
 'Price',
 'Customer ID',
 'Country']

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  str    
 1   StockCode    1067371 non-null  str    
 2   Description  1062989 non-null  str    
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  str    
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 65.1 MB


In [10]:
df.describe()

,Quantity,Price,Customer ID
count,1.067371e+06,1.067371e+06,824364.000000
mean,9.938898e+00,4.649388e+00,15324.638504
std,1.727058e+02,1.235531e+02,1697.464450
min,-8.099500e+04,-5.359436e+04,12346.000000
25%,1.000000e+00,1.250000e+00,13975.000000
50%,3.000000e+00,2.100000e+00,15255.000000
75%,1.000000e+01,4.150000e+00,16797.000000
max,8.099500e+04,3.897000e+04,18287.000000


In [11]:
missing = df.isnull().sum().sort_values(ascending=False)

missing

Customer ID    243007
Description      4382
StockCode           0
Invoice             0
Quantity            0
InvoiceDate         0
Price               0
Country             0
dtype: int64

In [16]:
missing_percent = (df.isnull().sum()/len(df))*100

missing_percent.sort_values(ascending=False)

Customer ID    22.766873
Description     0.410541
StockCode       0.000000
Invoice         0.000000
Quantity        0.000000
InvoiceDate     0.000000
Price           0.000000
Country         0.000000
dtype: float64

In [13]:
print("Duplicate Rows:", df.duplicated().sum())

Duplicate Rows: 34335


In [14]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [15]:
print("Start Date :", df["InvoiceDate"].min())
print("End Date   :", df["InvoiceDate"].max())

Start Date : 2009-12-01 07:45:00
End Date   : 2011-12-09 12:50:00


In [17]:
df["Customer ID"].nunique()

5942

In [18]:
df["StockCode"].nunique()

5305

In [19]:
df["Invoice"].nunique()

53628

In [20]:
df["Country"].nunique()

43

In [21]:
df["Revenue"] = df["Quantity"] * df["Price"]

In [22]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom,83.4
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom,100.8
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom,30.0


In [23]:
import os

os.makedirs("../data/processed", exist_ok=True)

df.to_csv(
    "../data/processed/raw_loaded_dataset.csv",
    index=False
)

In [24]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 9 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  str           
 1   StockCode    1067371 non-null  str           
 2   Description  1062989 non-null  str           
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[us]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  str           
 8   Revenue      1067371 non-null  float64       
dtypes: datetime64[us](1), float64(3), int64(1), str(4)
memory usage: 73.3 MB


In [25]:
missing

Customer ID    243007
Description      4382
StockCode           0
Invoice             0
Quantity            0
InvoiceDate         0
Price               0
Country             0
dtype: int64

In [26]:
df.duplicated().sum()

np.int64(34335)

In [27]:
print(df["InvoiceDate"].min())
print(df["InvoiceDate"].max())

2009-12-01 07:45:00
2011-12-09 12:50:00


## Observations

- The dataset contains **1,067,371 transaction records** spanning **December 2009 to December 2011**.
- There are **9 columns** describing customer purchases.
- Approximately **243,007 records (~22.8%)** have missing Customer IDs.
- **4,382 product descriptions** are missing.
- The dataset contains **34,335 duplicate records** that require further investigation.
- The dataset covers approximately **2 years** of customer transaction history, making it suitable for Cohort Analysis and CLTV modeling.

In [28]:
# Create a copy of the original dataset
clean_df = df.copy()

print(clean_df.shape)

(1067371, 9)


In [29]:
clean_df = clean_df.dropna(subset=["Customer ID"])

print(clean_df.shape)

(824364, 9)


In [30]:
clean_df = clean_df.drop_duplicates()

print(clean_df.shape)

(797885, 9)


In [31]:
clean_df = clean_df[
    ~clean_df["Invoice"].astype(str).str.startswith("C")
]

print(clean_df.shape)

(779495, 9)


In [32]:
clean_df = clean_df[
    clean_df["Quantity"] > 0
]

print(clean_df.shape)

(779495, 9)


In [33]:
clean_df = clean_df[
    clean_df["Price"] > 0
]

print(clean_df.shape)

(779425, 9)


In [34]:
clean_df.isnull().sum()

Invoice        0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
Price          0
Customer ID    0
Country        0
Revenue        0
dtype: int64

In [35]:
clean_df.to_csv(
    "../data/processed/cleaned_online_retail.csv",
    index=False
)

## Data Cleaning Summary

The dataset was cleaned using the following steps:

- Removed records with missing Customer IDs.
- Removed duplicate transactions.
- Removed cancelled invoices (Invoice starting with 'C').
- Removed transactions with non-positive quantities.
- Removed transactions with non-positive prices.

After cleaning, the dataset contains **779,425 valid customer transactions** and is ready for feature engineering.

In [36]:
clean_df["InvoiceMonth"] = clean_df["InvoiceDate"].dt.to_period("M")

In [37]:
clean_df["CohortMonth"] = (
    clean_df.groupby("Customer ID")["InvoiceDate"]
            .transform("min")
            .dt.to_period("M")
)

In [38]:
invoice_year = clean_df["InvoiceDate"].dt.year
invoice_month = clean_df["InvoiceDate"].dt.month

cohort_year = clean_df["CohortMonth"].dt.year
cohort_month = clean_df["CohortMonth"].dt.month

clean_df["CohortIndex"] = (
    (invoice_year - cohort_year) * 12
    + (invoice_month - cohort_month)
    + 1
)

In [39]:
clean_df["Year"] = clean_df["InvoiceDate"].dt.year
clean_df["Month"] = clean_df["InvoiceDate"].dt.month
clean_df["Day"] = clean_df["InvoiceDate"].dt.day
clean_df["DayOfWeek"] = clean_df["InvoiceDate"].dt.day_name()
clean_df["Hour"] = clean_df["InvoiceDate"].dt.hour

In [40]:
clean_df["CustomerRevenue"] = (
    clean_df.groupby("Customer ID")["Revenue"]
            .transform("sum")
)

In [41]:
clean_df["CustomerOrders"] = (
    clean_df.groupby("Customer ID")["Invoice"]
            .transform("nunique")
)

In [42]:
clean_df["AverageOrderValue"] = (
    clean_df["CustomerRevenue"] /
    clean_df["CustomerOrders"]
)

In [43]:
clean_df.to_csv(
    "../data/processed/feature_engineered_retail.csv",
    index=False
)